File: module_android/src/com/manateeworks/barcodescanner/MWOverlay.java
/*
 
 Version 1.0
 
 The MWOverlay class serves to greatly simplify the addition of a dynamic viewfinder (similar to the one implemented 
 in Manatee Works Barcode Scanners application) to your own application.
 
 Minimum setup assumes:
 1. Add MWOverlay.java to your project;
 2. Put MWOverlay.addOverlay(this, surfaceView); after initialization of surface view;
 3. Put MWOverlay.removeOverlay(); on closing the activity;
 
 If all three steps are done correctly, you should be able to see a default red viewfinder with a blinking line, 
 capable of updating itself automatically after changing any of the scanning parameters (scanning direction, scanning
 rectangles and active barcode symbologies).
  
 The appearance of the viewfinder and the blinking line can be further customized by changing colors, line width, 
 transparencies and similar, by setting the following properties:
 
 	MWOverlay.isViewportVisible;
	MWOverlay.isBlinkingLineVisible;
	MWOverlay.viewportLineWidth;
	MWOverlay.blinkingLineWidth;
	MWOverlay.viewportAlpha;
	MWOverlay.viewportLineAlpha;
	MWOverlay.blinkingLineAlpha;
	MWOverlay.blinkingSpeed;
	MWOverlay.viewportLineColor;
	MWOverlay.blinkingLineColor;

 */

package com.manateeworks.barcodescanner;

import java.util.Timer;
import java.util.TimerTask;

import com.manateeworks.BarcodeScanner;

import android.R.anim;
import android.content.Context;
import android.content.res.Configuration;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.RectF;
import android.util.DisplayMetrics;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;

public class MWOverlay extends View{
	
	public enum LayerType {
		   LT_VIEWPORT,
		   LT_LINE
		}
	
	private static boolean isAttached = false;
	public static boolean isViewportVisible = true;
	public static boolean isBlinkingLineVisible = true;

	public static float viewportLineWidth = 3.0f;
	public static float blinkingLineWidth = 1.0f;
	public static float viewportAlpha = 0.5f;
	public static float viewportLineAlpha = 0.5f;
	public static float blinkingLineAlpha = 1.0f;
	public static float blinkingSpeed = 0.25f;
	public static int viewportLineColor = 0xff0000;
	public static int blinkingLineColor = 0xff0000;

	private static int lastOrientation = -1;
	private static int lastMask = -1;
	private static float lastLeft = -1;
	private static float lastTop = -1;
	private static float lastWidth = -1;
	private static float lastHeight = -1;
	private static float lastBLinkingSpeed = -1;

	private static float dpiCorrection = 1;
	
	private static MWOverlay viewportLayer;
	private static MWOverlay lineLayer;
	
	
	private LayerType layerType;
	
	private static Timer  checkChangeTimer;
	
	public static MWOverlay addOverlay (Context context, View previewLayer) {
		
		isAttached = true;
		
		ViewGroup parent = (ViewGroup) previewLayer.getParent();
		
		DisplayMetrics metrics = context.getResources().getDisplayMetrics();
		dpiCorrection = metrics.density;
		
		viewportLayer = new MWOverlay(context);
		viewportLayer.layerType = LayerType.LT_VIEWPORT;
		
		lineLayer = new MWOverlay(context);
		lineLayer.layerType = LayerType.LT_LINE;
		
		
		viewportLayer.setDrawingCacheEnabled(true);
		lineLayer.setDrawingCacheEnabled(true);
		
		
		ViewGroup.LayoutParams rl= new ViewGroup.LayoutParams(ViewGroup.LayoutParams.FILL_PARENT, ViewGroup.LayoutParams.FILL_PARENT);
		   
		parent.addView(viewportLayer, rl);
		parent.addView(lineLayer, rl);
		int totalCount = parent.getChildCount();
		
		for (int i = 0; i < totalCount; i++){
			
			View child = parent.getChildAt(i);
			if (child.equals(previewLayer) || child.equals(viewportLayer) || child.equals(lineLayer)){
			
			} else {
				child.bringToFront();
				i--;
				totalCount--;
			}
			
		}
		
		checkChangeTimer = new Timer();
		checkChangeTimer.schedule(new TimerTask() {

			@Override
			public void run() {
				checkChange();
			}
		}, 200, 200);
		viewportLayer.postInvalidate();
		lineLayer.postInvalidate();
		
		updateAnimation();
		
		return viewportLayer;
		
	}
	
	public static void removeOverlay () {
		
		if (!isAttached)
			return;
		
		if (lineLayer == null || viewportLayer == null)
			return;
		
		checkChangeTimer.cancel();
		Animation animation = lineLayer.getAnimation();
		if (animation != null){
			animation.cancel();
			animation.reset();
		}
		ViewGroup viewParent = (ViewGroup) lineLayer.getParent();
		
		if (viewParent != null) {
			viewParent.removeView(lineLayer);
			viewParent.removeView(viewportLayer);
		}
		
		
		
	}
	
	private static void checkChange() {
		
		RectF frame = BarcodeScanner.MWBgetScanningRect(0);
		int orientation = BarcodeScanner.MWBgetDirection();
		
		if (orientation != lastOrientation || frame.left != lastLeft || frame.top != lastTop || frame.right != lastWidth || frame.bottom != lastHeight) {
			
			viewportLayer.postInvalidate();
			lineLayer.postInvalidate();
		}
		
		if (lastBLinkingSpeed != blinkingSpeed){
			updateAnimation();
		}
		
		if (isBlinkingLineVisible != (lineLayer.getVisibility() == View.VISIBLE)){
			
			lineLayer.postInvalidate();
		}
		
		if (isViewportVisible != (viewportLayer.getVisibility() == View.VISIBLE)){
			viewportLayer.postInvalidate();
		}
		
	}
	
	private static void updateAnimation() {
		
		AlphaAnimation animation = new AlphaAnimation(0, 1);
		animation.setRepeatMode(Animation.REVERSE);
		animation.setRepeatCount(Animation.INFINITE);
		animation.setDuration((long) (blinkingSpeed * 1000));
		
		lineLayer.startAnimation(animation);
		
		lastBLinkingSpeed = blinkingSpeed;
	}
	
	public MWOverlay(Context context) {
		super(context);
		// TODO Auto-generated constructor stub
	}
	
	
	 
	 
	 @Override
	 public void onDraw(Canvas canvas)
	 {
	        
		RectF frame = BarcodeScanner.MWBgetScanningRect(0);
		
		if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT){
			frame = new RectF(frame.top, frame.left, frame.bottom, frame.right);
		}
		
        lastLeft = frame.left;
        lastTop = frame.top;
        lastWidth = frame.right;
        lastHeight = frame.bottom;
		 
        int width = canvas.getWidth();
	    int height = canvas.getHeight();
	    
        float rectLeft =  frame.left  * width / 100.0f;
        float rectTop = frame.top  * height / 100.0f;
        float rectWidth = frame.right  * width / 100.0f;
        float rectHeight = frame.bottom  * height / 100.0f;
        
	    frame = new RectF(rectLeft, rectTop, rectWidth + rectLeft, rectHeight + rectTop);
	        
	    Paint paint = new Paint();
	    
	    if (layerType == LayerType.LT_VIEWPORT){
	    	
	    	if (isViewportVisible != (viewportLayer.getVisibility() == View.VISIBLE)){
				if (isViewportVisible){
					viewportLayer.setVisibility(View.VISIBLE);
				} else {
					viewportLayer.setVisibility(View.INVISIBLE);
				}
			}
	        
		    paint.setColor(0xff000000);
		    paint.setAlpha((int) (viewportAlpha * 255));
		    
		    
	        canvas.drawRect(0, 0, width, frame.top, paint);
	        canvas.drawRect(0, frame.top, frame.left, frame.bottom + 1, paint);
	        canvas.drawRect(frame.right + 1, frame.top, width, frame.bottom + 1, paint);
	        canvas.drawRect(0, frame.bottom + 1, width, height, paint);
	        
	        paint.setColor(viewportLineColor);
	        paint.setAlpha((int) (viewportLineAlpha * 255));
	        paint.setStyle(Paint.Style.STROKE);
	        paint.setStrokeWidth(viewportLineWidth * dpiCorrection);
	        
	        canvas.drawRect(frame.left, frame.top, frame.right, frame.bottom, paint);
	        
	    }	 else {
	    	
	    	if (isBlinkingLineVisible != (lineLayer.getVisibility() == View.VISIBLE)){
				if (isBlinkingLineVisible){
					lineLayer.setVisibility(View.VISIBLE);
					updateAnimation();
				} else {
					
					Animation animation = lineLayer.getAnimation();
					if (animation != null){
						animation.cancel();
						animation.reset();
					}
					lineLayer.setVisibility(View.INVISIBLE);
				}
			}
	    	
	        paint.setColor(blinkingLineColor);
	        paint.setStrokeWidth(blinkingLineWidth * dpiCorrection);
	        
	        long curTime = System.currentTimeMillis() % 10000000;
	        double position = ((double) curTime) / 1000.0d * 3.14d / blinkingSpeed;
	        
	        float lineAlpha = (float) (Math.abs(Math.sin(position)));
	        
	        paint.setAlpha((int) (blinkingLineAlpha * lineAlpha* 255));
	        paint.setAlpha((int) (blinkingLineAlpha * 255));
	        
	        int orientation = BarcodeScanner.MWBgetDirection();
	        
	        if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT){
	        	
	        	double pos1f = Math.log(BarcodeScanner.MWB_SCANDIRECTION_HORIZONTAL)/Math.log(2);
	        	double pos2f = Math.log(BarcodeScanner.MWB_SCANDIRECTION_VERTICAL)/Math.log(2);
	        	
	        	int pos1 = (int) (pos1f + 0.01);
	        	int pos2 = (int) (pos2f + 0.01);
	        	
	        	int bit1 = (orientation >> pos1) & 1;// bit at pos1
	        	int bit2 = (orientation >> pos2) & 1;// bit at pos2
	        	int mask = (bit2 << pos1) | (bit1 << pos2);
        	    orientation =  orientation & 0xc;
        	    orientation =  orientation | mask;
        	    
			}
	        
	        lastOrientation = orientation;
	        
	        if ((orientation & BarcodeScanner.MWB_SCANDIRECTION_HORIZONTAL) > 0 || (orientation & BarcodeScanner.MWB_SCANDIRECTION_OMNI) > 0){
	        
	        	float middle = frame.height() / 2 + frame.top;
	        	canvas.drawLine(frame.left, middle, frame.right, middle,  paint);
	        }
	        if ((orientation & BarcodeScanner.MWB_SCANDIRECTION_VERTICAL) > 0 || (orientation & BarcodeScanner.MWB_SCANDIRECTION_OMNI) > 0){
	            
	        	float middle = frame.width() / 2 + frame.left;
	        	canvas.drawLine(middle, frame.top, middle, frame.bottom - 1,  paint);
	        }
	        
	        if ((orientation & BarcodeScanner.MWB_SCANDIRECTION_OMNI) > 0){
	            
	        	canvas.drawLine(frame.left + 2, frame.top + 2, frame.right - 2, frame.bottom - 2, paint);
	        	canvas.drawLine(frame.right - 2, frame.top + 2, frame.left + 2, frame.bottom - 2, paint);
	        	
	        }
	    }

	}

}

File: module_ios/Classes/MWOverlay.m
//
//  MWOverlay.m
//  mobiscan_ALL
//  version 1.0
//
//  Created by vladimir zivkovic on 12/12/13.
//  Copyright (c) 2013 Manatee Works. All rights reserved.
//

/*
 
 Version 1.0
 
 The MWOverlay class serves to greatly simplify the addition of a dynamic viewfinder (similar to the one implemented in Manatee Works Barcode Scanners application) to your own application.
 Minimum setup assumes:
 1. Add MWOverlay.m and MWOverlay.h files to your project and import "MWOverlay.h" into the scanning view controller;
 2. Put [MWOverlay addToPreviewLayer:self.prevLayer]; after previewLayer is created and set;
 3. Put [MWOverlay removeFromPreviewLayer]; after closing the scanning session;
 If all three steps are done correctly, you should be able to see a default red viewfinder with a blinking line, capable of updating itself automatically after changing any of the scanning parameters (scanning direction, scanning rectangles and active barcode symbologies).
 The appearance of the viewfinder and the blinking line can be further customized by changing colors, line width, transparencies and similar, by using the following functions:
 + (void) setViewportVisible: (BOOL) value;
 + (void) setBlinkingLineVisible: (BOOL) value;
 + (void) setViewportLineWidth: (float) value;
 + (void) setBlinkingLineWidth: (float) value;
 + (void) setViewportAlpha: (float) value;
 + (void) setViewportLineAlpha: (float) value;
 + (void) setBlinkingLineAlpha: (float) value;
 + (void) setBlinkingSpeed: (float) value;
 + (void) setViewportLineRGBColor: (int) value;
 + (void) setBlinkingLineRGBColor: (int) value;
 + (void) setViewportLineUIColor: (UIColor*) value;
 + (void) setBlinkingLineUIColor: (UIColor*) value;

 */

#import "MWOverlay.h"
#import <AVFoundation/AVFoundation.h>
#import <CoreGraphics/CoreGraphics.h>
#import "BarcodeScanner.h"

#define CHANGE_TRACKING_INTERVAL 0.1

@implementation MWOverlay

CALayer *viewportLayer;
CALayer *lineLayer;
AVCaptureVideoPreviewLayer * previewLayer;
BOOL isAttached = NO;
BOOL isViewportVisible = YES;
BOOL isBlinkingLineVisible = YES;


MWOverlay *instance = nil;

float viewportLineWidth = 3.0;
float blinkingLineWidth = 1.0;
float viewportAlpha = 0.5;
float viewportLineAlpha = 0.5;
float blinkingLineAlpha = 1.0;
float blinkingSpeed = 0.25;
int viewportLineColor = 0xff0000;
int blinkingLineColor = 0xff0000;

int lastOrientation = -1;
int lastMask = -1;
float lastLeft = -1;
float lastTop = -1;
float lastWidth = -1;
float lastHeight = -1;




+ (void) addToPreviewLayer:(AVCaptureVideoPreviewLayer *) videoPreviewLayer
{
    viewportLayer = [[CALayer alloc] init];
    viewportLayer.frame = CGRectMake(0, 0, videoPreviewLayer.frame.size.width, videoPreviewLayer.frame.size.height);
    
    lineLayer = [[CALayer alloc] init];
    lineLayer.frame = CGRectMake(0, 0, videoPreviewLayer.frame.size.width, videoPreviewLayer.frame.size.height);
    
    [videoPreviewLayer addSublayer:viewportLayer];
    [videoPreviewLayer addSublayer:lineLayer];
    
    previewLayer = videoPreviewLayer;
    
    isAttached = YES;
    
    instance = [[MWOverlay alloc] init];
    
   
    [instance performSelector:@selector(checkForChanges) withObject:nil afterDelay:CHANGE_TRACKING_INTERVAL];
    [MWOverlay updateOverlay];
 
}

+ (void) removeFromPreviewLayer {
    
    if (!isAttached){
        return;
    }
    
    if (previewLayer){
        if (lineLayer){
            [lineLayer removeFromSuperlayer];
        }
        if (viewportLayer){
            [viewportLayer removeFromSuperlayer];
        }
    }
    
    isAttached = NO;
    
}

- (void) checkForChanges {
    
    if (isAttached){
        [instance performSelector:@selector(checkForChanges) withObject:nil afterDelay:CHANGE_TRACKING_INTERVAL];
    } else {
        return;
    }
    
    float left, top, width, height;
    int res = MWB_getScanningRect(0, &left, &top, &width, &height);
    
    if (res == 0){
    
        int orientation = MWB_getDirection();
        
        if (orientation != lastOrientation || left != lastLeft || top != lastTop || width != lastWidth || height != lastHeight) {
            
            [instance performSelectorOnMainThread:@selector(updateOverlayMainThread) withObject:nil waitUntilDone:NO];
            NSLog(@"Change detected");
            
        }
        
        lastOrientation = orientation;
        lastLeft = left;
        lastTop = top;
        lastWidth = width;
        lastHeight = height;
        
    }
    
    
}

- (void) updateOverlayMainThread {
    
    [MWOverlay updateOverlay];
    
}

+ (void) updateOverlay{
    
    if (!isAttached || !previewLayer){
        return;
    }
    
    float yScale = 1.0;
    float yOffset = 0.0;
    float xScale = 1.0;
    float xOffset = 0.0;
    
    //aspect ratio correction available only on ios 6+
    if ([previewLayer respondsToSelector:@selector(captureDevicePointOfInterestForPoint:)]){
    
        CGPoint p1 = [previewLayer captureDevicePointOfInterestForPoint:CGPointMake(0,0)];
        
        yScale = 1.0/(1 + (p1.y - 1)*2);
        yOffset = (1.0 - yScale) / 2.0 * 100;
        
        xScale = 1.0/(1 + (p1.x - 1)*2);
        xOffset = (1.0 - xScale) / 2.0 * 100;
        
        
        if (previewLayer.connection.videoOrientation == AVCaptureVideoOrientationPortrait || previewLayer.connection.videoOrientation == AVCaptureVideoOrientationPortraitUpsideDown){
            
            yScale = 1.0/(1 + (p1.x - 1)*2);
            yOffset = (1.0 - yScale) / 2.0 * 100;
            
            xScale = 1.0/(1 + (p1.y - 1)*2);
            xOffset = (1.0 - xScale) / 2.0 * 100;
            
        }
    }
    
    
    viewportLayer.hidden = !isViewportVisible;
    lineLayer.hidden = !isBlinkingLineVisible;
    
   
    int overlayWidth = viewportLayer.frame.size.width;
    int overlayHeight = viewportLayer.frame.size.height;
    
    CGRect cgRect = viewportLayer.frame;
    
    
    UIGraphicsBeginImageContext(cgRect.size);
    CGContextRef context = UIGraphicsGetCurrentContext();
	UIGraphicsPushContext(context);
    
    
    
    float left, top, width, height;
    
    MWB_getScanningRect(0, &left, &top, &width, &height);
    
    if (previewLayer.connection.videoOrientation == AVCaptureVideoOrientationPortrait || previewLayer.connection.videoOrientation == AVCaptureVideoOrientationPortraitUpsideDown){
        
        float tmp = left;
        left = top;
        top = tmp;
        tmp = height;
        height = width;
        width = tmp;
        
    }
    
    CGRect rect = CGRectMake(xOffset + left * xScale, yOffset + top * yScale, width * xScale, height * yScale);
    
    rect.origin.x *= overlayWidth;
    rect.origin.x /= 100.0;
    rect.origin.y *= overlayHeight;
    rect.origin.y /= 100.0;
    rect.size.width *= overlayWidth;
    rect.size.width /= 100.0;
    rect.size.height *= overlayHeight;
    rect.size.height /= 100.0;
    
    CGContextSetRGBFillColor(context, 0, 0, 0, viewportAlpha);
    CGContextFillRect(context, CGRectMake(0,0,overlayWidth,overlayHeight));
    CGContextClearRect(context, rect);
    
    float r = (viewportLineColor >> 16) / 255.0;
    float g = ((viewportLineColor & 0x00ff00) >> 8) / 255.0;
    float b = (viewportLineColor & 0x0000ff) / 255.0;
    
    
    CGContextSetRGBStrokeColor(context, r, g, b, 0.5);
    CGContextStrokeRectWithWidth(context, rect, viewportLineWidth);
    
    UIGraphicsPopContext();
    
    viewportLayer.contents = (id)[UIGraphicsGetImageFromCurrentImageContext() CGImage];
    
  
    
    CGContextClearRect(context, cgRect);
    
    r = (blinkingLineColor >> 16) / 255.0;
    g = ((blinkingLineColor & 0x00ff00) >> 8) / 255.0;
    b = (blinkingLineColor & 0x0000ff) / 255.0;

    CGContextSetRGBStrokeColor(context, r, g, b, 1);
    
    int orientation = MWB_getDirection();
    
    if (previewLayer.connection.videoOrientation == AVCaptureVideoOrientationPortrait || previewLayer.connection.videoOrientation == AVCaptureVideoOrientationPortraitUpsideDown){
        
        double pos1f = log(MWB_SCANDIRECTION_HORIZONTAL) / log(2);
        double pos2f = log(MWB_SCANDIRECTION_VERTICAL) / log(2);
        
        int pos1 = (int)(pos1f + 0.01);
        int pos2 = (int)(pos2f + 0.01);
        
        int bit1 = (orientation >> pos1) & 1;// bit at pos1
        int bit2 = (orientation >> pos2) & 1;// bit at pos2
        int mask = (bit2 << pos1) | (bit1 << pos2);
        orientation = orientation & 0xc;
        orientation = orientation | mask;
        
    }

    
    if (orientation & MWB_SCANDIRECTION_HORIZONTAL || orientation & MWB_SCANDIRECTION_OMNI || orientation & MWB_SCANDIRECTION_AUTODETECT){
        CGContextSetLineWidth(context, blinkingLineWidth);
        CGContextMoveToPoint(context, rect.origin.x, rect.origin.y + rect.size.height / 2);
        CGContextAddLineToPoint(context, rect.origin.x + rect.size.width, rect.origin.y + rect.size.height / 2);
        CGContextStrokePath(context);
    }
    
    if (orientation & MWB_SCANDIRECTION_VERTICAL || orientation & MWB_SCANDIRECTION_OMNI || orientation & MWB_SCANDIRECTION_AUTODETECT){
       
        CGContextMoveToPoint(context, rect.origin.x + rect.size.width / 2, rect.origin.y);
        CGContextAddLineToPoint(context, rect.origin.x + rect.size.width / 2, rect.origin.y + rect.size.height);
        CGContextStrokePath(context);
    }
    
    if (orientation & MWB_SCANDIRECTION_OMNI || orientation & MWB_SCANDIRECTION_AUTODETECT){
        CGContextMoveToPoint(context, rect.origin.x , rect.origin.y);
        CGContextAddLineToPoint(context, rect.origin.x + rect.size.width , rect.origin.y + rect.size.height);
        CGContextStrokePath(context);
        
        CGContextMoveToPoint(context, rect.origin.x + rect.size.width, rect.origin.y);
        CGContextAddLineToPoint(context, rect.origin.x , rect.origin.y + rect.size.height);
        CGContextStrokePath(context);
    }
    
    lineLayer.contents = (id)[UIGraphicsGetImageFromCurrentImageContext() CGImage];
    
    UIGraphicsEndImageContext();
    
   [MWOverlay startLineAnimation];
    
    
}


+ (void) startLineAnimation {
    [lineLayer removeAllAnimations];
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"opacity"];
    [animation setFromValue:[NSNumber numberWithFloat:blinkingLineAlpha]];
    [animation setToValue:[NSNumber numberWithFloat:0.0]];
    [animation setDuration:blinkingSpeed];
    [animation setTimingFunction:[CAMediaTimingFunction
                                  functionWithName:kCAMediaTimingFunctionLinear]];
    [animation setAutoreverses:YES];
    [animation setRepeatCount:INFINITY];
    [lineLayer addAnimation:animation forKey:@"opacity"];
}

+ (void) setViewportVisible: (BOOL) value {
    
    isViewportVisible = value;
    
}

+ (void) setBlinkingLineVisible: (BOOL) value {
    
    isBlinkingLineVisible = value;
    
}

+ (void) setViewportLineWidth: (float) value {
    
    viewportLineWidth = value;
    [MWOverlay updateOverlay];
    
}

+ (void) setBlinkingLineWidth: (float) value {
    
    blinkingLineWidth = value;
    [MWOverlay updateOverlay];
    
}

+ (void) setViewportAlpha: (float) value {
    
    viewportAlpha = value;
    [MWOverlay updateOverlay];
    
}

+ (void) setViewportLineAlpha: (float) value {
    
    viewportLineAlpha = value;
    [MWOverlay updateOverlay];
    
}

+ (void) setBlinkingLineAlpha: (float) value {
    
    blinkingLineAlpha = value;
    [MWOverlay updateOverlay];
    
}

+ (void) setBlinkingSpeed: (float) value {
    
    blinkingSpeed = value;
    [MWOverlay updateOverlay];
    
}

+ (void) setViewportLineRGBColor: (int) value {
    
    viewportLineColor = value;
    [MWOverlay updateOverlay];
    
}

+ (void) setBlinkingLineRGBColor: (int) value {
    
    blinkingLineColor = value;
    [MWOverlay updateOverlay];
    
}

+ (void) setViewportLineUIColor: (UIColor*) value {
    
    CGColorRef color = [value CGColor];
    
    int numComponents = CGColorGetNumberOfComponents(color);
    
    if (numComponents >= 3)
    {
        const CGFloat *components = CGColorGetComponents(color);
        CGFloat red = components[0];
        CGFloat green = components[1];
        CGFloat blue = components[2];
    
        int intColor = (((int) (red * 255)) << 16) + (((int) (green * 255)) << 8) + ((int) (blue * 255)) ;
        
        viewportLineColor = intColor;
        [MWOverlay updateOverlay];
        
    }
}

+ (void) setBlinkingLineUIColor: (UIColor*) value {
    
    CGColorRef color = [value CGColor];
    
    int numComponents = CGColorGetNumberOfComponents(color);
    
    if (numComponents >= 3)
    {
        const CGFloat *components = CGColorGetComponents(color);
        CGFloat red = components[0];
        CGFloat green = components[1];
        CGFloat blue = components[2];
        
        int intColor = (((int) (red * 255)) << 16) + (((int) (green * 255)) << 8) + ((int) (blue * 255)) ;
        
        blinkingLineColor = intColor;
        [MWOverlay updateOverlay];
        
    }
}




@end
